src/nlp/Preprocessing.py

```diff
@@ -0,0 +1,477 @@
+      have multple words associated. In such case the full word that makes more sense given the
+      context is chosen (see findOptimalOption method)
+      @ In, abbrDatabase, pandas dataframe, dataframe containing library of abbreviations
+                                            and their correspoding full expression
+      @ In, text, str, string of text that will be analyzed
+      @ In, type, string, type of abbreviation method ('spellcheck','hard','mixed') that are employed
+                          to determine which words are abbreviations that nned to be expanded
+                          * spellcheck: in this case spellchecker is used to identify words that
+                                        are not recognized
+                          * hard: here we directly search for the abbreviations in the provided
+                                  sentence
+                          * mixed: here we perform first a "hard" search followed by a "spellcheck"
+                                   search
+      @ Out, options, list, list of corrected text options
+    """
+    abbreviationSet = set(abbrDatabase['Abbreviation'].values)
+    if type == 'spellcheck':
+      unknowns = self.getMisspelledWords(text)
+    elif type == 'hard' or type=='mixed':
+      unknowns = []
+      splitSent = text.split()
+      for word in splitSent:
+        if word.lower() in abbreviationSet:
+          unknowns.append(word)
+      if type=='mixed':
+        set1 = set(self.getMisspelledWords(text))
+        set2 = set(unknowns)
+        unknowns = list(set1.union(set2))
+
+    corrections={}
+    for word in unknowns:
+      if word.lower() in abbrDatabase['Abbreviation'].values:
+        locs = list(abbrDatabase['Abbreviation'][abbrDatabase['Abbreviation']==word].index.values)
+        if locs:
+          corrections[word] = abbrDatabase['Full'][locs].values.tolist()
+      else:
+        # Here we are addressing the fact that the abbreviation database will never be complete
+        # Given an abbreviation that is not part of the abbreviation database, we are looking for a
+        # a subset of abbreviations the abbreviation database that are close enough (and consider
+        # them as possible candidates
+        from difflib import SequenceMatcher
+        corrections[word] = []
+        abbreviationDS = abbrDatabase['Abbreviation'].values
+        for index,abbr in enumerate(abbreviationDS):
+          if SequenceMatcher(None, word, abbr).ratio()>0.8:
+            corrections[word].append(abbrDatabase['Full'].values.tolist()[index])
+      if not corrections[word]:
+        corrections.pop(word)
+
+    combinations = list(itertools.product(*list(corrections.values())))
+    options = []
+    for comb in combinations:
+      corrected = text
+      for index,key in enumerate(corrections.keys()):
+        corrected = re.sub(r"\b%s\b" % str(key) , comb[index], corrected)
```

Comment:
This can fail for certain symbols in key (such as "**" or "(").  On my local copy I added:
```
        try:
          corrected = re.sub(r"\b%s\b" % str(key) , comb[index], corrected)
        except re.error as e:
          print(f"Failed on {key} for {e}")
```
but I am not sure the best way to fix this.